File: LeetCode/131分割回文串.go
var ans [][]string
var path []string

func partition(s string) [][]string {
	ans = [][]string{}
	path = []string{}
	backtrack(0, s)
	return ans
}

func backtrack(ind int, s string) {
	if ind >= len(s) {
		temp := make([]string, len(path))
		copy(temp, path)
		ans = append(ans, temp)
		return
	}
	for i := ind; i < len(s); i++ {
		if isValid(s, ind, i) {
			str := s[ind : i+1]
			path = append(path, str)
		} else {
			continue
		}
		backtrack(i+1, s)
		path = path[:len(path)-1]
	}
	return
}

func isValid(s string, left int, right int) bool {
	i, j := left, right
	for i < j {
		if s[i] != s[j] {
			return false
		}
		i++
		j--
	}
	return true
}

File: LeetCode/131分割回文串.java
class Solution {
    List<List<String>> ans; 
    LinkedList<String> path;
    public List<List<String>> partition(String s) {
        ans = new ArrayList<>();
        path = new LinkedList<>();
        backtracking(0, s);
        return ans;
    }

    public void backtracking(int ind, String s) {
        if (ind >= s.length()) {
            ans.add(new ArrayList<>(path));
            return ;
        }
        for (int i = ind; i < s.length(); i++) {
            if (isValid(s, ind, i)) {
                String str = s.substring(ind, i + 1);
                path.add(str);
            } else continue;
            backtracking(i + 1, s);
            path.removeLast();
        }
        return ;
    }

    public boolean isValid(String s, int left, int right) {
        for (int i = left, j = right; i < j; i++, j--) {
            if (s.charAt(i) != s.charAt(j)) return false;
        }
        return true;
    }
}